### PR TITLE
Add support for pre-hashing passwords in proxysql_mysql_users module

### PR DESCRIPTION
**Pull request by mjbnz**
##### SUMMARY

If user passwords are converted to mysql hashed format using this sequence:
```sql
LOAD MYSQL USERS TO RUNTIME;
SAVE MYSQL USERS FROM RUNTIME;
SAVE MYSQL USERS TO DISK;
```
Then any subsequent calls to the `proxysql_mysql_users` module will treat the user as needing an update to change the password back to cleartext.

This PR simply adds a check for the password provided in mysql hashed form.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

`proxysql_mysql_users`

##### ADDITIONAL INFORMATION
The query `WHERE` clause to check the password is changed from:
```sql
 AND password = <cleartext>
```
to:
```sql
 AND password in (<cleartext>, <hashed>)
```

### CODE REVIEW

**Comment by mjbnz** [comment]
Helps if I actually add the file to the index before amending the commit :|

**Comment by bmildren** [comment]
Hi @mjbnz - that's an interesting point, thanks for bringing this up!

We supply an encrypted pw to ProxySQL so this isn't something we've come across (we also tend to use memory as the source of truth, so typically only save to disk/load to runtime).

Mulling this over, I think it might introduce an unintended consequence (albeit probably minor) of not being able to update an existing cleartext pw in memory to its encrypted equivalent via the module (other than doing the load/save to runtime dance).  This could also cause some further edge cases when ProxySQL starts to support the caching_sha2_password plugin.

I wonder if another route might be to expand this further, and rather than defer this to ProxySQL to encrypt the pw (using save from runtime), add an explicit option to the module to allow a supplied cleartext password to be encrypted (we could maybe add another option to support a default of mysql native password encryption for now, but leave it open to support new encryption methods as/when they're added).

Definitely happy to discuss this further!

**Comment by mjbnz** [comment]
Aha - that's also a good point (changing from the same password as clear to hashed in memory being ignored). Agreed, it's very minor consequence, but still one that should be avoided.

We use ansible-vault to store the cleartext passwords for database users - it should probably be the hash, but the vault is our authoritative source of truth for mysql passwords. I personally don't like storing hashes, as there is potential for them to get out of sync with the cleartext version, whereever that ends up being stored. That said, cleartext versions should definitely be store securely!

So, in summary, I think I like your idea better of adding a flag to ask the module to hash the password before handing it to proxysql. I'd be happy to implement that, but unsure about the best way to implement the encryption method support option you propose.

**Comment by bmildren** [comment]
Maybe something along the lines of this:

```python
...
encrypt_password=dict(default=False, type='bool'),
encryption_method=dict(default='mysql_native_password', choices=['mysql_native_password']),
...


...
def mysql_native_password(cleartext_password):
    mysql_native_encrypted_password = "*" + sha1(sha1(cleartext_password).digest()).hexdigest().upper()
    return mysql_native_encrypted_password


def encrypt_cleartext_password(password_to_encrypt, encryption_method):
    encrypted_password = encryption_method(password_to_encrypt)
    return encrypted_password
...


...
    encryption_method_mapper = {'mysql_native_password': mysql_native_password}
    authentication_plugin = encryption_method_mapper[encryption_method]

    if encrypt_password:
        encrypted_password = encrypt_cleartext_password(password, authentication_plugin)
...
```

**Comment by mjbnz** [comment]
> Maybe something along the lines of this:

Yep, that makes perfect sense - force push above to change to an implementation of that.

File: plugins/modules/proxysql_mysql_users.py [hunk 1]
```diff
@@ -395,6 +412,8 @@ def main():
             config_file=dict(default='', type='path'),
             username=dict(required=True, type='str'),
             password=dict(no_log=True, type='str'),
+            encrypt_password=dict(default=False, type='bool'),
+            encryption_method=dict(default='mysql_native_password', choices=['mysql_native_password']),
```
Comment (mjbnz) [inline plugins/modules/proxysql_mysql_users.py:416]: Can that `choices=` option be passed the keys of `encryption_method_map` ?

**Comment by mjbnz** [comment]
CI still failed after fixing the problems it reported... and it's reporting the same problems. Is this a result of doing a force push ?

**Comment by mjbnz** [comment]
Let's try a rebase! \o/

File: plugins/modules/proxysql_mysql_users.py (hunk 1 above)
Comment (mjbnz) [reply plugins/modules/proxysql_mysql_users.py]: Turns out, yes!

File: plugins/modules/proxysql_mysql_users.py [hunk 2]
```diff
@@ -177,6 +190,21 @@ def load_config_to_runtime(cursor):
     return True
 
 
+def mysql_native_password(cleartext_password):
+    mysql_native_encrypted_password = "*" + sha1(sha1(cleartext_password.encode('utf-8')).digest()).hexdigest().upper()
```
Comment (bmildren) [inline plugins/modules/proxysql_mysql_users.py:194]: I think we can import `to_bytes` from `from ansible.module_utils._text` (alongside `to_native`), and use that instead of .encode here.

**Comment by bmildren** [comment]
just one comment about using `to_bytes` rather than .encode other than that lgtm!

File: plugins/modules/proxysql_mysql_users.py [hunk 3]
```diff
@@ -177,6 +190,21 @@ def load_config_to_runtime(cursor):
     return True
 
 
+def mysql_native_password(cleartext_password):
```
Comment (bmildren) [inline plugins/modules/proxysql_mysql_users.py:193]: just a nitpick, but we could also rename this function to `_mysql_native_password` to indicate it shouldn't be called directly (it should always be called via `encrypt_cleartext_password`)

Comment (mjbnz) [reply plugins/modules/proxysql_mysql_users.py]: Yes, I agree. Done.

File: plugins/modules/proxysql_mysql_users.py (hunk 2 above)
Comment (mjbnz) [reply plugins/modules/proxysql_mysql_users.py]: Ahh, yes, I wasn't aware of `to_bytes()`. Done.... however, I realise I've messed it up, I'll wait till CI finishes and fails, then fix it.

Comment (mjbnz) [reply plugins/modules/proxysql_mysql_users.py]: Now fixed.

**Comment by bmildren** [comment]
awesome lgtm!
